frame/pallet-xvm/src/evm.rs

```diff
@@ -1,61 +1,67 @@
 //! EVM support for XVM pallet.
 
 use crate::*;
+use pallet_evm::GasWeightMapping;
 use sp_core::{H160, U256};
 use sp_runtime::traits::Get;
 
-pub struct EVM<I, T, C>(sp_std::marker::PhantomData<(I, T, C)>);
+/// EVM adapter for XVM calls.
+///
+/// This adapter supports generic XVM calls and encode it into EVM native calls
+/// using Solidity ABI codec (https://docs.soliditylang.org/en/v0.8.16/abi-spec.html).
+pub struct EVM<I, T>(sp_std::marker::PhantomData<(I, T)>);
 
-impl<VmId, I, T, C> SyncVM<VmId, T::AccountId> for EVM<I, T, C>
+impl<I, T> SyncVM<T::AccountId> for EVM<I, T>
 where
     I: Get<VmId>,
     T: pallet_evm::Config + frame_system::Config,
-    C: XvmCodec,
 {
     fn id() -> VmId {
         I::get()
     }
 
-    fn xvm_call(
-        _context: XvmContext<VmId>,
-        from: T::AccountId,
-        to: Vec<u8>,
-        input: Vec<u8>,
-        metadata: Vec<u8>,
-    ) -> Result<Vec<u8>, Vec<u8>> {
+    fn xvm_call(context: XvmContext, from: T::AccountId, to: Vec<u8>, input: Vec<u8>) -> XvmResult {
         log::trace!(
             target: "xvm::EVM::xvm_call",
-            "Start EVM XVM: {:?}, {:?}, {:?}, {:?}",
-            from, to, input, metadata,
+            "Start EVM XVM: {:?}, {:?}, {:?}",
+            from, to, input,
         );
-        let data = C::convert(input, metadata)?;
         let value = U256::from(0u64);
         let max_fee_per_gas = U256::from(3450898690u64);
-        let gas_limit = 4000000u64;
-        let _nonce = frame_system::Pallet::<T>::account(from.clone()).nonce;
-        let evm_to: H160 = Decode::decode(&mut to.as_ref())
-            .map_err(|_| b"`to` argument decode failure".to_vec())?;
+        let gas_limit = T::GasWeightMapping::weight_to_gas(context.max_weight);
+        log::trace!(
+            target: "xvm::EVM::xvm_call",
+            "EVM xvm call gas limit: {:?} or as weight: {:?}", gas_limit, context.max_weight);
+        let evm_to = Decode::decode(&mut to.as_ref()).map_err(|_| XvmCallError {
+            error: XvmError::EncodingFailure,
+            consumed_weight: PLACEHOLDER_WEIGHT,
+        })?;
 
         let res = pallet_evm::Pallet::<T>::call(
             frame_support::dispatch::RawOrigin::Root.into(),
             H160::from_slice(&from.encode()[0..20]),
             evm_to,
-            data,
+            input,
             value,
             gas_limit,
             max_fee_per_gas,
             None,
             None,
             Vec::new(),
-        );
+        )
+        .map_err(|_e| XvmCallError {
+            error: XvmError::ExecutionError(Vec::default()), // TODO: make error mapping make more sense
+            consumed_weight: 42u64, // TODO: res.actual_weight.map(|x| x.ref_time()).unwrap_or(context.max_weight),
```

Comment:
We should go with worst case scenario here and specify `consumed_weight: context.max_weight`

Reply:
Yes, but it not works with new weighs, so, I just hack it to make it works.

Reply:
Can you please explain more? What doesn't work?

Reply:
I believe problem here is different types for `Weight` and context types. We should rework it using new substrate weight system.